scripts/reqcheck/main.sh

```diff
@@ -0,0 +1,29 @@
+#!/bin/bash -eu
```

Comment:
We're overthinking this. The extra infrastructure makes it significantly harder to understand what's going on.

This should do the job:
```sh
#!/bin/bash -eu

function err() {
  echo "$@" >&2
  return 1
}

function main() {
  declare -ri node_version=$(expr "$(node -v)" : 'v\([^.]*\)')
  (( node_version >= 12 )) || err "Need node version >= 12. Found ${node_version}"

  declare -ri npm_version=$(expr "$(npm -v)" : '\([^.]*\)')
  (( npm_version >= 7 )) || err "Need npm version >= 7. Found ${npm_version}"
}

main
```

Let's call the script `check_requirements.sh`. No need to abbreviate. It's not something we will be running often.

I'd like the user to call check_requirements directly, not via npm run, in case npm is not present.

Let's drop the setup script. There's no much point in setting up only if nvm is available, and it's not installing nvm itself.

My suggestion is:
- Use the simple script above.
- Add a line to the requirements section of the readme saying you can run `./scripts/check_requirements.sh` to check the requirements.
- Add the script call to preinstall

Reply:
I have tried to build the script with what you said in mind, such that we can extend it down the line more easily, adding different categories of things to check for like we discussed (android, etc). Right now you can check javascript dependencies by passing the `-j` flag, but other flags could be provided later for other groups of dependencies. Just want to confirm you no longer want that functionality.

My issue with the script you have generously provided is that it will diverge from the package.json as the source of truth as we update our node version going forward, as you've hard coded the values. This will cause confusion down the line. It also doesn't respect minor versions (less important).

Really, all I care about is having a core source of truth that we can respect. It doesn't have to be the `package.json` - I am happy to make a `requirements.txt` or something that's easier to understand, and remove node and npm from the `engines` in the package.json so it's clearer.

Perhaps I did it incorrectly, but the intent with calling to nvm in setup was to just switch to the correct version of node if nvm was present for the dev's convenience. `nvm use` looks for an `.nvmrc` and uses that version. It does nothing if nvm is not present, I was just trying to be helpful.

I called it `reqcheck` so it remains consistent with our other checks, e.g. `shellcheck` - is it okay if I rename `shellcheck` to `check_shell`?

Lastly, I can add it to preinstall, but if you watch the video I recorded above, you'll see that `preinstall` is only run as the results of `npm install` are finalized. `node_modules` are still added, and files still written to; I feel `preinstall` gives us a false sense of security and doesn't give us much.

Reply:
Let's establish there's no right or wrong here. :-) We are talking about a tradeoff.

This is adding a lot of code to understand, break and maintain, which is a concern of mine. The amount of code is not proportional to the importance of the problem either. Furthermore, I'd claim that my suggestion is a lot easier to extend, since you can just add a new line or two in the simpler script, without having to understand anything else, and it works with or without a package.json.

It's useful to leave the door open to extensibility in the future, but we should focus on the use case we have. Currently we are just trying to fix npm engine check for new developers that don't read the readme. (But if they don't read the readme they won't even know about `npm run setup`).

I just want a quick fix that will let us move forward with the npm migration (a much more consequential change) and won't add significant overhead.

I hear you about having a single source of truth. In practice I'm ok with having that duplicated in multiple places for now because we rarely change it. Notice that we already duplicate it anyway in the CI config and you're duplicating it in the nvmrc. *If* that turns out to be a real problem, then we can address it. As you make progress with the feature work, you will see we have much bigger fish to fry.

I'm actually comfortable with the strict engines check and the recommendation to use `npm install` (rather than `npm ci`). That should enforce the restriction, right? It may not work with older npm versions, but that's fine. People can figure it out.